builtin/v8/miner/invariants.go

```diff
@@ -0,0 +1,879 @@
+package miner
+
+import (
+	addr "github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/builtin/v8/util"
+
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
+)
+
+type DealSummary struct {
+	SectorStart      abi.ChainEpoch
+	SectorExpiration abi.ChainEpoch
+}
+
+type StateSummary struct {
+	LivePower           PowerPair
+	ActivePower         PowerPair
+	FaultyPower         PowerPair
+	Deals               map[abi.DealID]DealSummary
+	WindowPoStProofType abi.RegisteredPoStProof
+	DeadlineCronActive  bool
+}
+
+// Checks internal invariants of init state.
+func CheckStateInvariants(st *State, store adt.Store, balance abi.TokenAmount) (*StateSummary, *builtin.MessageAccumulator) {
+	acc := &builtin.MessageAccumulator{}
+	sectorSize := abi.SectorSize(0)
+	minerSummary := &StateSummary{
+		LivePower:           NewPowerPairZero(),
+		ActivePower:         NewPowerPairZero(),
+		FaultyPower:         NewPowerPairZero(),
+		WindowPoStProofType: 0,
+		DeadlineCronActive:  st.DeadlineCronActive,
+	}
+
+	// Load data from linked structures.
+	if info, err := st.GetInfo(store); err != nil {
+		acc.Addf("error loading miner info: %v", err)
+		// Stop here, it's too hard to make other useful checks.
+		return minerSummary, acc
+	} else {
+		minerSummary.WindowPoStProofType = info.WindowPoStProofType
+		sectorSize = info.SectorSize
+		CheckMinerInfo(info, acc)
+	}
+
+	CheckMinerBalances(st, store, balance, acc)
+
+	var allocatedSectors bitfield.BitField
+	var allocatedSectorsMap map[uint64]bool
+	if err := store.Get(store.Context(), st.AllocatedSectors, &allocatedSectors); err != nil {
+		acc.Addf("error loading allocated sector bitfield: %v", err)
+	} else {
+		allocatedSectorsMap, err = allocatedSectors.AllMap(1 << 30)
```

Comment:
Is there a reason we're using 2^30 here? We have miners with sector numbers greater than that, which is the source of some of the mainnet failures (but it's not invalid to have such large sector numbers, the max is 2^64).

Reply:
That's what was in the specs-actors version. Please change if it's wrong :)

Reply:
Done!